backend/logic/order/index.js

```diff
@@ -203,8 +211,141 @@ async function processNewOrder({
   return order
 }
 
+const validPaymentStateTransitions = {
+  [OrderPaymentStatuses.Refunded]: [],
+  [OrderPaymentStatuses.Rejected]: [],
+  [OrderPaymentStatuses.Pending]: [
+    OrderPaymentStatuses.Paid,
+    OrderPaymentStatuses.Rejected,
+    OrderPaymentStatuses.Refunded
+  ],
+  [OrderPaymentStatuses.Paid]: [OrderPaymentStatuses.Refunded]
+}
+
+/**
+ * Returns the validity of the payment state
+ * transition on an order
+ * @param {model.Order} order
+ * @param {enums.OrderPaymentStatuses} newState
+ *
+ * @returns {Boolean} true if valid
+ */
+const isValidTransition = function (order, newState) {
+  return get(
+    validPaymentStateTransitions,
+    order.paymentStatus,
+    validPaymentStateTransitions[OrderPaymentStatuses.Pending]
+  ).includes(newState)
+}
+
+/**
+ * Updates the payment state of an order
+ * @param {model.Order} order
+ * @param {enums.OrderPaymentStatuses} newState
+ * @param {mode.Shop} shop
+ *
+ * @returns {{
+ *  success {Boolean}
+ *  reason {String|null} error message if any
+ * }}
+ */
+async function updatePaymentStatus(order, newState, shop) {
+  if (order.paymentStatus === newState) {
+    // No change, Ignore
+    return { success: true }
+  }
+
+  if (!isValidTransition(order, newState)) {
+    return {
+      reason: `Cannot change payment state from ${order.paymentStatus} to ${newState}`
+    }
+  }
+
+  let refundError = null
+  if (newState === OrderPaymentStatuses.Refunded) {
+    // Initiate a refund in case of Stripe and PayPal
+    switch (order.paymentType) {
+      case OrderPaymentTypes.CreditCard:
+        refundError = await processStripeRefund({ shop, order })
+        break
+      case OrderPaymentTypes.PayPal:
+        refundError = await processPayPalRefund({ shop, order })
+        break
+    }
+  } else if (newState === OrderPaymentStatuses.Paid) {
+    const shopConfig = getConfig(shop.config)
+    if (shopConfig.printful && shopConfig.printfulAutoFulfill) {
+      await autoFulfillOrder(order, shopConfig, shop)
+    }
+  }
+
+  await order.update({
+    paymentStatus: newState,
+    data: {
+      ...order.data,
+      refundError
```

Comment:
nit: this will include refundError in the data even for non-refund related updates.
Maybe only include refundError conditionally upon an update to Refunded state?